## The Competition

Robots are smart… by design. To fully understand and properly navigate a task, however, they need input about their environment.

In this competition, you’ll help robots recognize the floor surface they’re standing on using data collected from Inertial Measurement Units (IMU sensors).

We’ve collected IMU sensor data while driving a small mobile robot over different floor surfaces on the university premises. The task is to predict which one of the nine floor types (carpet, tiles, concrete) the robot is on using sensor data such as acceleration and velocity. Succeed and you'll help improve the navigation of robots without assistance across many different surfaces, so they won’t fall down on the job.

### Libraries

In [1]:
import pandas as pd

### Load & check out the data

In [2]:
# Load sensor data
X_train = pd.read_csv('data/X_train.csv')

In [15]:
X_train.head(135)

row_id  series_id  measurement_number  orientation_X  orientation_Y  \
0      0_0          0                   0       -0.75853       -0.63435   
1      0_1          0                   1       -0.75853       -0.63434   
2      0_2          0                   2       -0.75853       -0.63435   
3      0_3          0                   3       -0.75852       -0.63436   
4      0_4          0                   4       -0.75852       -0.63435   
5      0_5          0                   5       -0.75853       -0.63439   
6      0_6          0                   6       -0.75853       -0.63441   
7      0_7          0                   7       -0.75852       -0.63444   
8      0_8          0                   8       -0.75851       -0.63445   
9      0_9          0                   9       -0.75851       -0.63443   
10    0_10          0                  10       -0.75848       -0.63443   
11    0_11          0                  11       -0.75847       -0.63444   
12    0_12          0                  12       -0.75846       -0.63448   
13    0_13          0                  13       -0.75846       -0.63445   
14    0_14          0                  14       -0.75842       -0.63447   
15    0_15          0                  15       -0.75838       -0.63447   
16    0_16          0                  16       -0.75835       -0.63447   
17    0_17          0                  17       -0.75835       -0.63451   
18    0_18          0                  18       -0.75835       -0.63453   
19    0_19          0                  19       -0.75833       -0.63456   
20    0_20          0                  20       -0.75830       -0.63456   
21    0_21          0                  21       -0.75828       -0.63453   
22    0_22          0                  22       -0.75826       -0.63453   
23    0_23          0                  23       -0.75825       -0.63453   
24    0_24          0                  24       -0.75824       -0.63454   
25    0_25          0                  25       -0.75824       -0.63450   
26    0_26          0                  26       -0.75823       -0.63450   
27    0_27          0                  27       -0.75822       -0.63449   
28    0_28          0                  28       -0.75822       -0.63448   
29    0_29          0                  29       -0.75823       -0.63446   
..     ...        ...                 ...            ...            ...   
105  0_105          0                 105       -0.75906       -0.63338   
106  0_106          0                 106       -0.75906       -0.63338   
107  0_107          0                 107       -0.75906       -0.63339   
108  0_108          0                 108       -0.75908       -0.63339   
109  0_109          0                 109       -0.75909       -0.63337   
110  0_110          0                 110       -0.75910       -0.63337   
111  0_111          0                 111       -0.75911       -0.63336   
112  0_112          0                 112       -0.75913       -0.63336   
113  0_113          0                 113       -0.75917       -0.63331   
114  0_114          0                 114       -0.75919       -0.63331   
115  0_115          0                 115       -0.75921       -0.63330   
116  0_116          0                 116       -0.75924       -0.63329   
117  0_117          0                 117       -0.75927       -0.63327   
118  0_118          0                 118       -0.75930       -0.63324   
119  0_119          0                 119       -0.75931       -0.63323   
120  0_120          0                 120       -0.75933       -0.63321   
121  0_121          0                 121       -0.75938       -0.63318   
122  0_122          0                 122       -0.75942       -0.63317   
123  0_123          0                 123       -0.75943       -0.63316   
124  0_124          0                 124       -0.75945       -0.63313   
125  0_125          0                 125       -0.75949       -0.63309   
126  0_126          0        

In [12]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 487680 entries, 0 to 487679
Data columns (total 13 columns):
row_id                   487680 non-null object
series_id                487680 non-null int64
measurement_number       487680 non-null int64
orientation_X            487680 non-null float64
orientation_Y            487680 non-null float64
orientation_Z            487680 non-null float64
orientation_W            487680 non-null float64
angular_velocity_X       487680 non-null float64
angular_velocity_Y       487680 non-null float64
angular_velocity_Z       487680 non-null float64
linear_acceleration_X    487680 non-null float64
linear_acceleration_Y    487680 non-null float64
linear_acceleration_Z    487680 non-null float64
dtypes: float64(10), int64(2), object(1)
memory usage: 48.4+ MB


**X_[train/test].csv** - the input data, covering 10 sensor channels and 128 measurements per time series plus three ID columns:

* row_id: The ID for this row.
* series_id: ID number for the measurement series. Foreign key to y_train/sample_submission.
* measurement_number: Measurement number within the series.

The **10 sensor channels** are:

The **orientation** channels encode the current angles how the robot is oriented as a quaternion
* orientation_X
* orientation_Y
* orientation_Z
* orientation_W

**Angular velocity** describes the angle and speed of motion
* angular_velocity_X
* angular_velocity_Y
* angular_velocity_Z

**Linear acceleration** components describe how the speed is changing at different times
* linear_acceleration_X
* linear_acceleration_Y
* linear_acceleration_Z

**_Note_**  
In the context of relational databases, a **foreign key** is a field (or collection of fields) in one table that uniquely identifies a row of another table or the same table.  
Source [Wikipedia - Foreign key](https://en.wikipedia.org/wiki/Foreign_key)

In [6]:
# Load label data
y_train = pd.read_csv('data/y_train.csv')

In [7]:
y_train.head()

series_id  group_id        surface
0          0        13  fine_concrete
1          1        31       concrete
2          2        20       concrete
3          3        31       concrete
4          4        22     soft_tiles

In [13]:
y_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3810 entries, 0 to 3809
Data columns (total 3 columns):
series_id    3810 non-null int64
group_id     3810 non-null int64
surface      3810 non-null object
dtypes: int64(2), object(1)
memory usage: 89.4+ KB


**y_train.csv** - the surfaces for training set.

* series_id: ID number for the measurement series.
* group_id: ID number for all of the measurements taken in a recording session. Provided for the training set only, to enable more cross validation strategies.
* surface: the target for this competition.

### Understanding the Data

#### Simplified Problem Description

A scientist walks a robot every day for a cetain period of time on different floor types.  
During each "walk" different kinds of sensor data is collected (The data comes from 10 sensors).  
Each "walk" is $n$ steps (time intervals $\Delta t$) long where after each interval each sensor records a datum.  
The scientist walks the robot for $d$ days and after each day annotates (labels) the collected sensor data set with the corresponding floor type.

So our scientist should have $n * d$ data points in total.  

Each "walk" can be described as a Time Series.  
We know from the description of the dataset that

* $n = 128$ from **X_train**
* $d = 3810$ from **y_train**

So the number of total data points should be

$$
n * d = 128 * 3810 = 487,680
$$

which is the number of data point in **X_train**.

In [9]:
sample_submission = pd.read_csv('data/sample_submission.csv')

In [10]:
sample_submission.head()

series_id   surface
0          0  concrete
1          1  concrete
2          2  concrete
3          3  concrete
4          4  concrete

In [11]:
sample_submission.tail()

series_id   surface
3811       3811  concrete
3812       3812  concrete
3813       3813  concrete
3814       3814  concrete
3815       3815  concrete

### Questions to answer

### EDA

### Solution Approach

* What kind of problem?
* Which solutions could be used?